backends/p4tools/modules/testgen/core/small_step/expr_stepper.h

```diff
@@ -18,6 +17,149 @@ namespace P4Tools::P4Testgen {
+        /// @param args the list of arguments being passed to method.
+        /// @param state the state in which the call is being made, with the call at the top of the
+        ///     current continuation body.
+        ///
+        /// @returns true if a matching extern-method implementation was found; false otherwise.
+        ///
+        /// A BUG occurs if the receiver is not an extern or if the call is ambiguous.
+        std::optional<MethodImpl> find(const IR::PathExpression &externObjectRef,
+                                       const IR::ID &methodName,
+                                       const IR::Vector<IR::Argument> &args) const {
+            // We have to check the extern type here. We may receive a specialized canonical type,
+            // which we need to unpack.
+            const IR::Type_Extern *externType = nullptr;
+            if (const auto *type = externObjectRef.type->to<IR::Type_Extern>()) {
+                externType = type;
+            } else if (const auto *specType =
+                           externObjectRef.type->to<IR::Type_SpecializedCanonical>()) {
+                CHECK_NULL(specType->substituted);
+                externType = specType->substituted->checkedTo<IR::Type_Extern>();
+            } else if (externObjectRef.path->name == IR::ID("*method")) {
+            } else {
+                BUG("Not a valid extern: %1% with member %2%. Type is %3%.", externObjectRef,
+                    methodName, externObjectRef.type->node_type_name());
+            }
+
+            cstring qualifiedMethodName = externType->name + "." + methodName;
+            if (impls.count(qualifiedMethodName) == 0) {
+                return std::nullopt;
+            }
+
+            const auto &submap = impls.at(qualifiedMethodName);
+            if (submap.count(args.size()) == 0) {
+                return std::nullopt;
+            }
+
+            // Find matching methods: if any arguments are named, then the parameter name must
+            // match.
+            std::optional<MethodImpl> matchingImpl;
+            for (const auto &pair : submap.at(args.size())) {
+                const auto &paramNames = pair.first;
+                const auto &methodImpl = pair.second;
+
+                if (matches(paramNames, args)) {
+                    BUG_CHECK(!matchingImpl, "Ambiguous extern method call: %1%",
+                              qualifiedMethodName);
+                    matchingImpl = methodImpl;
+                }
+            }
+
+            return matchingImpl;
+        }
+
+     private:
+        /// A two-level map. First-level keys are method names qualified by the extern type name
+        /// (e.g., "packet_in.advance". Second-level keys are the number of arguments. Values are
+        /// lists of matching extern-method implementations, paired with the names of the method
+        /// parameters.
+        std::map<cstring, std::map<uint, std::list<std::pair<std::vector<cstring>, MethodImpl>>>>
```

Comment:
It would be probably better to replace the `list` by `vector`, `list` is rarely better unless we really want stable references and insertion in the middle.